lib/yarp/ffi.rb

```diff
@@ -9,135 +12,200 @@ module YARP
-    def self.pointer(size)
-      pointer = FFI::MemoryPointer.new(size)
+    # Yields a yp_string_t pointer to the given block.
+    def self.with_string(filepath, &block)
+      string = FFI::MemoryPointer.new(SIZEOF_YP_STRING)
+
       begin
-        yield pointer
+        raise unless yp_string_mapped_init(string, filepath)
+        yield YPString.new(string)
       ensure
-        pointer.free
+        yp_string_free(string)
+        string.free
       end
     end
-
-    def self.yp_string(&block)
-      pointer(LibRubyParser::SIZEOF_YP_STRING, &block)
-    end
   end
+
+  # Mark the LibRubyParser module as private as it should only be called through
+  # the YARP module.
   private_constant :LibRubyParser
 
-  VERSION = LibRubyParser.yp_version.to_s
+  # The version constant is set by reading the result of calling yp_version.
+  VERSION = LibRubyParser.yp_version.read_string
 
   def self.dump_internal(source, source_size, filepath)
-    buffer = LibRubyParser::Buffer.new
-    begin
-      raise unless LibRubyParser.yp_buffer_init(buffer)
-      metadata = nil
-      if filepath
-        metadata = [filepath.bytesize].pack('L') + filepath + [0].pack('L')
-      end
+    LibRubyParser.with_buffer do |buffer|
+      metadata = [filepath.bytesize, filepath.b, 0].pack("LA*L") if filepath
       LibRubyParser.yp_parse_serialize(source, source_size, buffer, metadata)
       buffer.to_ruby_string
-    ensure
-      LibRubyParser.yp_buffer_free(buffer)
-      buffer.pointer.free
     end
   end
   private_class_method :dump_internal
 
+  # Mirror the YARP.dump API by using the serialization API.
   def self.dump(code, filepath = nil)
     dump_internal(code, code.bytesize, filepath)
   end
 
+  # Mirror the YARP.dump_file API by using the serialization API.
   def self.dump_file(filepath)
-    LibRubyParser.yp_string do |contents|
-      raise unless LibRubyParser.yp_string_mapped_init(contents, filepath)
-      dump_internal(LibRubyParser.yp_string_source(contents), LibRubyParser.yp_string_length(contents), filepath)
-    ensure
-      LibRubyParser.yp_string_free(contents)
+    LibRubyParser.with_string(filepath) do |string|
+      dump_internal(string.source, string.length, filepath)
     end
   end
 
+  # Mirror the YARP.lex API by using the serialization API.
   def self.lex(code, filepath = nil)
-    buffer = LibRubyParser::Buffer.new
-    begin
-      raise unless LibRubyParser.yp_buffer_init(buffer)
+    LibRubyParser.with_buffer do |buffer|
       LibRubyParser.yp_lex_serialize(code, code.bytesize, filepath, buffer)
-      serialized = buffer.to_ruby_string
 
       source = Source.new(code)
-      parse_result = YARP::Serialize.load_tokens(source, serialized)
-
-      ParseResult.new(parse_result.value, parse_result.comments, parse_result.errors, parse_result.warnings, source)
-    ensure
-      LibRubyParser.yp_buffer_free(buffer)
-      buffer.pointer.free
+      Serialize.load_tokens(source, buffer.to_ruby_string).with_source(source)
```

Comment:
Actually we shouldn't need the `with_source` here, it should be already correct.
I'll make a PR.